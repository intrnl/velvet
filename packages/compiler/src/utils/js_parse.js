import * as acorn from 'acorn';
import { walk } from 'estree-walker';
import { generate } from 'astring';

/**
 * @param {string} source
 * @returns {import('estree').Program}
 */
export function parse (source) {
	/** @type {import('estree').Comment[]} */
	let comments = [];

	let program = acorn.parse(source, {
		onComment: comments,
		ecmaVersion: 'latest',
		sourceType: 'module',
	});

	reattach_comments(program);
	return program;
}

export function parse_expression (source, position = 0) {
	/** @type {import('estree').Comment[]} */
	let comments = [];

	let node = acorn.parseExpressionAt(source, position, {
		onComment: comments,
		ecmaVersion: 'latest',
		sourceType: 'module',
	});

	reattach_comments(node, comments);
	return node;
}

/**
 * @param {import('estree').Node} ast
 * @param {*} map
 * @returns {string}
 */
export function print (ast, map) {
	return generate(ast, {
		comments: true,
		sourceMap: map,
	});
}

function reattach_comments (ast, comments) {
	walk(ast, {
		/** @param {import('estree').Node} node */
		enter (node) {
			/** @type {import('estree').Comment} */
			let comment;
			let leading_comments = node.leadingComments ||= [];

			while ((comment = comments[0]) && comment.start < node.start) {
				leading_comments.push(comments.shift());
			}
		},
		/** @param {import('estree').Node} node */
		leave (node) {
			let trailing_comments = node.trailingComments ||= [];

			if (comments[0]) {
				let slice = source.slice(node.end, comments[0].start);

				if (/^[,) \t]*$/.test(slice)) {
					trailing_comments.push(comments.shift());
				}
			}
		},
	});
}
