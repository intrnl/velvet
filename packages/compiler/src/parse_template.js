import * as p from './utils/template_parser.js';
import * as t from './utils/template_types.js';
import * as j from './utils/js_parse.js';
import { is_void, closing_tag_omitted } from './utils/html.js';


export function parse_template (content) {
	let state = p.create_state(content);

	let parser = _parse_fragment;

	while (state.index < state.content.length) {
		parser = parser(state) || _parse_fragment;
	}

	if (state.stack.length > 1) {
		throw p.error(state, 'unexpected end of input');
	}

	/** @type {t.Fragment} */
	let root = p.current(state);

	root.start = root.children[0]?.start || 0;
	root.end = root.children[root.children.length - 1]?.end || 0;

	return root;
}

/**
 * @param {p.ParserState} state
 */
function _parse_fragment (state) {
	let char = state.content[state.index];

	if (char === '<') {
		return _parse_element;
	}

	if (char === '{') {
		return _parse_expression;
	}

	return _parse_text;
}

/**
 * @param {p.ParserState} state
 */
function _parse_text (state) {
	let start = state.index;
	let raw = '';

	for (; state.index < state.content.length; state.index++) {
		let char = state.content[state.index];

		if (char === '<' || char === '{') {
			break;
		}

		raw += char;
	}

	let node = t.text(raw);
	node.start = start;
	node.end = state.index;

	p.current(state).children.push(node);
}

/**
 * @param {p.ParserState} state
 */
function _parse_expression (state) {
	let start = state.index;

	p.eat(state, '{', 'opening expression bracket');

	// opening logic block
	if (p.eat(state, '#')) {
		if (p.eat(state, 'if')) {
			p.eat_whitespace(state, true);

			let test = _read_expression(state);

			p.eat_whitespace(state);
			p.eat(state, '}', 'closing #if bracket');

			let consequent = t.fragment();
			let node = t.conditional_statement(test, consequent);
			node.start = start;

			p.current(state).children.push(node);
			p.push(state, node, consequent);
			return;
		}

		if (p.eat(state, 'each')) {
			p.eat_whitespace(state, true);

			let local = _read_expression(state);

			if (!_is_identifier(local)) {
				throw p.error(state, 'expected an assignment', local.start);
			}

			p.eat_whitespace(state, true);
			p.eat(state, 'of', true);
			p.eat_whitespace(state, true);

			let expression = _read_expression(state);

			p.eat_whitespace(state);
			p.eat(state, '}', 'closing #each bracket');

			let body = t.fragment();
			let node = t.loop_statement(expression, local, body);
			node.start = start;

			p.current(state).children.push(node);
			p.push(state, node, body);
			return;
		}

		if (p.eat(state, 'await')) {
			p.eat_whitespace(state, true);

			let argument = _read_expression(state);

			p.eat_whitespace(state);

			let to_resolve = p.eat(state, 'then');
			let to_reject = !to_resolve && p.eat(state, 'catch');
			p.eat_whitespace(state);

			let local = null;

			if ((to_resolve || to_reject) && !p.match(state, '}')) {
				local = _read_expression(state);

				if (!_is_identifier(local)) {
					throw p.error(state, 'expected an assignment', local.start);
				}

				p.eat_whitespace(state);
			}

			p.eat(state, '}', 'closing #await bracket');

			let block = t.fragment();
			let clause = t.await_clause(local, block);

			let node = t.await_statement(argument, !to_resolve && !to_reject ? block : null, to_resolve ? clause : null, to_reject ? clause : null);
			node.start = start;

			p.current(state).children.push(node);
			p.push(state, node, block);
			return;
		}

		throw p.error(state, 'unknown logic block opening');
	}

	// closing logic block
	if (p.eat(state, '/')) {
		let statement = p.current(state, 1);
		let block = p.current(state);
		let expected;

		if (!statement) {
			throw p.error(state, 'unexpected logic block closing');
		}

		if (statement.type === 'ConditionalStatement') {
			expected = 'if';
		}
		else if (statement.type === 'LoopStatement') {
			expected = 'each';
		}
		else if (statement.type === 'AwaitStatement') {
			expected = 'await';
		}
		else {
			throw p.error(state, 'unexpected logic block closing');
		}

		p.eat(state, expected, `closing /${expected} block`);
		p.eat_whitespace(state);
		p.eat(state, '}', `closing /${expected} bracket`);

		block.start = block.children[0]?.start || statement.start;
		block.end = block.children[block.children.length - 1]?.end || statement.start;
		statement.end = state.index;

		if (statement.type === 'ConditionalStatement') {
			// 0 = Fragment
			// 1 = ConditionalStatement
			// 2 = ConditionalStatement?
			while (p.current(state, 1) === p.current(state, 2)?.alternate) {
				let statement = p.current(state, 2);
				statement.end = state.index;

				p.pop(state, 1);
			}
		}

		p.pop(state, 2);
		return;
	}

	// intermediary logic
	if (p.eat(state, ':else')) {
		let statement = p.current(state, 1);
		let prev_block = p.current(state);

		if (!statement) {
			throw p.error(state, 'unexpected usage of :else outside of #if and #each');
		}

		prev_block.start = prev_block.children[0]?.start || statement.start;
		prev_block.end = prev_block.children[prev_block.children.length - 1]?.end || statement.start;

		if (statement.type === 'ConditionalStatement') {
			let additional = p.eat_whitespace(state) && p.eat(state, 'if');

			if (additional) {
				p.eat_whitespace(state, true);

				let test = _read_expression(state);
				p.eat_whitespace(state);

				let block = t.fragment();
				let node = t.conditional_statement(test, block);
				node.start = start;

				statement.alternate = node;

				p.pop(state, 1);
				p.push(state, node, block);
			}
			else if (statement.alternate) {
				throw p.error(state, 'only one :else can be defined');
			}
			else {
				let block = t.fragment();

				statement.alternate = block;

				p.pop(state, 1);
				p.push(state, block);
			}

			p.eat(state, '}', 'closing :else bracket');
			return;
		}

		if (statement.type === 'LoopStatement') {
			if (statement.alternate) {
				throw p.error(state, 'only one :else can be defined');
			}

			p.eat_whitespace(state);

			let block = t.fragment();

			statement.alternate = block;

			p.pop(state, 1);
			p.push(state, block);

			p.eat(state, '}', 'closing :else bracket');
			return;
		}

		throw p.error(state, 'unexpected usage of :else outside of #if and #each');
	}

	if (p.eat(state, ':then')) {
		let statement = p.current(state, 1);

		if (statement?.type !== 'AwaitStatement') {
			throw p.error(state, 'unexpected usage of :then outside of #await');
		}

		if (statement.resolved) {
			throw p.error(state, 'only one :then can be defined');
		}

		if (statement.rejected) {
			throw p.error(state, 'cannot have :then after :catch');
		}

		let has_whitespace = p.eat_whitespace(state);
		let has_expression = !p.eat(state, '}');
		let expression = null;

		if (has_expression) {
			if (!has_whitespace) {
				throw p.error(state, 'expected whitespace');
			}

			expression = _read_expression(state);

			if (!_is_identifier(expression)) {
				throw p.error(state, 'expected an assignment', expression.start);
			}

			p.eat_whitespace(state);
			p.eat(state, '}', 'closing :then bracket');
		}

		let block = t.fragment();
		let node = t.await_clause(expression, block);

		statement.resolved = node;

		p.pop(state, 1);
		p.push(state, block);
		return;
	}

	if (p.eat(state, ':catch')) {
		let statement = p.current(state, 1);

		if (statement?.type !== 'AwaitStatement') {
			throw p.error(state, 'unexpected usage of :catch outside of #await');
		}

		if (statement.rejected) {
			throw p.error(state, 'only one :catch can be defined');
		}

		let has_whitespace = p.eat_whitespace(state);
		let has_expression = !p.eat(state, '}');
		let expression = null;

		if (has_expression) {
			if (!has_whitespace) {
				throw p.error(state, 'expected whitespace');
			}

			expression = _read_expression(state);

			if (!_is_identifier(expression)) {
				throw p.error(state, 'expected an assignment', expression.start);
			}

			p.eat_whitespace(state);
			p.eat(state, '}', 'closing :catch bracket');
		}

		let block = t.fragment();
		let node = t.await_clause(expression, block);

		statement.rejected = node;

		p.pop(state, 1);
		p.push(state, block);
		return;
	}

	// named expression
	let id = null;

	if (p.eat(state, '@')) {
		id = _read_expression(state);

		if (id.type !== 'Identifier') {
			throw p.error(state, 'expected an identifier', id.start);
		}

		p.eat_whitespace(state, true);
	}

	p.eat_whitespace(state);
	let expression = _read_expression(state);
	p.eat_whitespace(state);

	p.eat(state, '}', 'closing expression bracket');

	let node = t.expression(expression, id);

	node.start = start;
	node.end = state.index;

	p.current(state).children.push(node);
}

/**
 * @param {p.ParserState} state
 */
function _parse_element (state) {
	let start = state.index;
	let parent = p.current(state);

	p.eat(state, '<', 'opening tag bracket');

	// comment
	if (p.eat(state, '!--')) {
		let data = p.eat_until(state, /-->/g);
		p.eat(state, '-->', 'closing comment');

		let node = t.comment(data);
		node.start = start;
		node.end = state.index;

		parent.children.push(node);
		return;
	}

	let is_closing = p.eat(state, '/');
	let name = p.eat_until(state, /[\s>]/g);

	if (name === 'v:self') {
		// check for recursion
		let legal = false;

		for (let index = state.stack.length - 1; index >= 0; index--) {
			let node = state.stack[index];

			if (
				(node.type === 'Element' && node.component) ||
				node.type === 'ConditionalStatement' ||
				node.type === 'LoopStatement'
			) {
				legal = true;
				break;
			}
		}

		if (!legal) {
			throw p.error(state, 'v:self placement causes infinite recursion');
		}
	}
	else if (name === 'v:component') {
		// left blank
	}
	else if (name[0] === 'v' && name[1] === ':') {
		throw p.error(state, 'unknown special element');
	}

	if (is_closing) {
		if (is_void(name)) {
			throw p.error(state, `${tag} is a void element and cannot have a closing tag or children`);
		}

		p.eat_whitespace(state);
		p.eat(state, '>', 'closing tag bracket');

		while (parent.name !== name) {
			if (parent.type !== 'Element') {
				let message = state._last_auto_closed?.name === name
					? `</${name}> attempted to close <${name}> that were already closed by <${state._last_auto_closed.tag}>`
					: `</${name}> attempted to close an element that was not open`;

				throw p.error(state, message, start);
			}

			parent.end = start;
			p.pop(state);

			parent = p.current();
		}

		parent.end = state.index;
		p.pop(state);

		if (state._last_auto_closed && state.stack.length < state._last_auto_closed.depth) {
			state._last_auto_closed = null;
		}

		return;
	}
	else if (closing_tag_omitted(parent.name, name)) {
		parent.end = start;
		p.pop(state);

		state._last_auto_closed = {
			name: parent.name,
			tag: name,
			depth: state.stack.length,
		};
	}

	// attributes parsing
	let attributes = [];
	let attribute_names = new Set();

	if (p.eat_whitespace(state)) {
		while (state.content[state.index] !== '>') {
			let start = state.index;

			if (p.eat(state, '{')) {
				p.eat_whitespace(state);
				p.eat(state, '...', 'spread pattern');

				let expression = _read_expression(state);

				p.eat_whitespace(state);
				p.eat(state, '}', 'closing spread bracket');

				let node = t.attribute_spread(expression);
				node.start = start;
				node.end = state.index;

				attributes.push(node);
				continue;
			}

			let attr_name = p.eat_until(state, /[\s=\/>'"]/g);
			let attr_value = null;

			if (!attr_name) {
				break;
			}

			if (attribute_names.has(attr_name)) {
				throw p.error(state, `duplicate ${attr_name} attribute`, start);
			}

			let end = state.index;

			p.eat_whitespace(state);

			if (p.eat_pattern(state, /['"]/g)) {
				throw p.error(state, 'expected attribute assignment');
			}

			if (p.eat(state, '=')) {
				let value_start = state.index;
				let quotation = p.eat_pattern(state, /['"]/g);

				if (quotation) {
					let end_pattern = quotation === '"' ? /"/g : /'/g;
					let data = p.eat_until(state, end_pattern);
					p.eat_pattern(state, end_pattern, 'closing quotation mark');

					let node = t.text(data);
					node.start = value_start;
					node.end = state.index;

					attr_value = node;
				}
				else if (p.eat(state, '{')) {
					p.eat_whitespace(state);

					let expression = _read_expression(state);

					p.eat_whitespace(state);
					p.eat(state, '}', 'closing expression bracket');

					let node = t.expression(expression);
					node.start = value_start;
					node.end = state.index;

					attr_value = node;
				}
				else {
					let data = p.eat_until(state, /[\s=\/>'"]/g);

					let node = t.text(data);
					node.start = value_start;
					node.end = state.index;

					attr_value = node;
				}

				end = state.index;
				p.eat_whitespace(state);
			}

			let node = t.attribute(attr_name, attr_value);
			node.start = start;
			node.end = end;

			attributes.push(node);
			attribute_names.add(attr_name);
		}
	}

	let self_closing = p.eat(state, '/') || is_void(name);
	p.eat(state, '>', 'closing tag bracket');

	let node = t.element(name, self_closing, attributes);
	node.start = start;
	node.end = state.index;

	if (name === 'script' || name === 'style') {
		// we shouldn't parse into script and style elements
		let pattern = name === 'script' ? /<\/script\s*>/g : /<\/style\s*>/g;
		let data = p.eat_until(state, pattern);

		p.eat_pattern(state, pattern, `${name} closing tag`);

		let text = t.text(data);
		let node = t.element(name, [], [text]);

		parent.children.push(node);
		return;
	}

	parent.children.push(node);

	if (!self_closing) {
		p.push(state, node);
	}
}

/**
 * @param {p.ParserState} state
 */
function _read_expression (state) {
	try {
		let node = j.parse_expression(state.content, state.index);

		let parens = 0;

		for (let index = state.index; index < node.start; index++) {
			if (state.content[index] === '(') {
				parens++;
			}
		}

		let index = node.end;

		while (parens > 0) {
			let char = state.content[index];

			if (char === ')') {
				parens--;
			}
			else if (!p.is_whitespace(char)) {
				throw p.error(state, `expected closing parenthesis`, index);
			}

			index++;
		}

		state.index = index;
		return node;
	}
	catch (error) {
		throw p.error(state, `JS parsing error: ${error.message}`);
	}
}

/**
 * @param {import('estree').Expression} node
 * @returns {node is import('estree').Identifier}
 */
function _is_identifier (node) {
	return node.type === 'Identifier';
}
