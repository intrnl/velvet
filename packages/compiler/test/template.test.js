import { describe, it, expect } from 'vitest';

import { parse_template } from '../src/parse_template.js';
import { transform_template } from '../src/transform_template.js';
import { print } from '../src/utils/js_parse.js';


describe('attribute', () => {
	it('static', () => {
		let template = `<div class='foo'></div>`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});

	it('static unquoted', () => {
		let template = `<div class=foo></div>`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});

	it('dynamic', () => {
		let template = `<div class={className}></div>`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});

	it('static boolean', () => {
		let template = `<textarea readonly></textarea>`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});

	it('dynamic boolean', () => {
		let template = `<textarea ?readonly={is_readonly}></textarea>`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});

	it('dynamic boolean no value', () => {
		let template = `<textarea ?readonly></textarea>`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});

	it('property', () => {
		let template = `<input .value={value}>`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});

	it('property no value', () => {
		let template = `<input .value>`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});
});

describe('component', () => {
	it('v:self on a custom element', () => {
		let template = `<x-app><v:self>hello world!</v:self></x-app>`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});
});

describe('expression', () => {
	it('custom element', () => {
		let template = `<x-app>hello {name}!</x-app>`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});

	it('component', () => {
		let template = `<Main>hello {name}!</Main>`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});
});

describe('conditional logic', () => {
	it('consequent', () => {
		let template = `{#if foo}<div>foo!</div>{/if}`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});

	it('consequent and alternate', () => {
		let template = `{#if foo}<div>foo!</div>{:else}<div>bar!</div>{/if}`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});

	it('consequent and alternate test', () => {
		let template = `{#if foo}<div>foo!</div>{:else if bar}<div>bar!</div>{/if}`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});

	it('nested consequent', () => {
		let template = `{#if foo}<div>foo!</div>{#if bar}<div>bar!</div>{/if}{/if}`;

		let fragment = parse_template(template);
		expect(fragment).toMatchSnapshot();

		let program = transform_template(fragment);

		expect(print(program)).toMatchSnapshot();
	});
});
