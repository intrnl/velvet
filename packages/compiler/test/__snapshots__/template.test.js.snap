// Vitest Snapshot v1

exports[`attribute > dynamic 1`] = `
{
  "children": [
    {
      "attributes": [
        {
          "end": 22,
          "name": "class",
          "start": 5,
          "type": "Attribute",
          "value": {
            "end": 22,
            "expression": Node {
              "comments": [],
              "end": 21,
              "name": "className",
              "path": {
                "parent": undefined,
              },
              "start": 12,
              "trailingComments": [],
              "type": "Identifier",
            },
            "id": null,
            "start": 11,
            "type": "Expression",
          },
        },
      ],
      "children": [],
      "component": false,
      "end": 29,
      "inline": false,
      "name": "div",
      "self_closing": false,
      "start": 0,
      "type": "Element",
    },
  ],
  "end": 29,
  "start": 0,
  "type": "Fragment",
}
`;

exports[`attribute > dynamic 2`] = `
"let %%template0 = @html(\\"<div></div>\\");
let %fragment0 = @clone(%template0);
$: @attr(%child0, \\"class\\", className);
@append(%fragment0, $$root);
"
`;

exports[`attribute > dynamic boolean 1`] = `
{
  "children": [
    {
      "attributes": [
        {
          "end": 33,
          "name": "?readonly",
          "start": 10,
          "type": "Attribute",
          "value": {
            "end": 33,
            "expression": Node {
              "comments": [],
              "end": 32,
              "name": "is_readonly",
              "path": {
                "parent": undefined,
              },
              "start": 21,
              "trailingComments": [],
              "type": "Identifier",
            },
            "id": null,
            "start": 20,
            "type": "Expression",
          },
        },
      ],
      "children": [],
      "component": false,
      "end": 45,
      "inline": false,
      "name": "textarea",
      "self_closing": false,
      "start": 0,
      "type": "Element",
    },
  ],
  "end": 45,
  "start": 0,
  "type": "Fragment",
}
`;

exports[`attribute > dynamic boolean 2`] = `
"let %%template0 = @html(\\"<textarea></textarea>\\");
let %fragment0 = @clone(%template0);
let %child0 = @traverse(%fragment0, [0]);
$: @toggle(%child0, \\"readonly\\", is_readonly);
@append(%fragment0, $$root);
"
`;

exports[`attribute > static 1`] = `
{
  "children": [
    {
      "attributes": [
        {
          "end": 16,
          "name": "class",
          "start": 5,
          "type": "Attribute",
          "value": {
            "decoded": "foo",
            "end": 16,
            "start": 11,
            "type": "Text",
            "value": "foo",
          },
        },
      ],
      "children": [],
      "component": false,
      "end": 23,
      "inline": false,
      "name": "div",
      "self_closing": false,
      "start": 0,
      "type": "Element",
    },
  ],
  "end": 23,
  "start": 0,
  "type": "Fragment",
}
`;

exports[`attribute > static 2`] = `
"let %%template0 = @html(\\"<div class=\\\\\\"foo\\\\\\"></div>\\");
let %fragment0 = @clone(%template0);
@append(%fragment0, $$root);
"
`;

exports[`attribute > static boolean 1`] = `
{
  "children": [
    {
      "attributes": [
        {
          "end": 18,
          "name": "readonly",
          "start": 10,
          "type": "Attribute",
          "value": null,
        },
      ],
      "children": [],
      "component": false,
      "end": 30,
      "inline": false,
      "name": "textarea",
      "self_closing": false,
      "start": 0,
      "type": "Element",
    },
  ],
  "end": 30,
  "start": 0,
  "type": "Fragment",
}
`;

exports[`attribute > static boolean 2`] = `
"let %%template0 = @html(\\"<textarea readonly></textarea>\\");
let %fragment0 = @clone(%template0);
@append(%fragment0, $$root);
"
`;

exports[`attribute > static unquoted 1`] = `
{
  "children": [
    {
      "attributes": [
        {
          "end": 14,
          "name": "class",
          "start": 5,
          "type": "Attribute",
          "value": {
            "decoded": "foo",
            "end": 14,
            "start": 11,
            "type": "Text",
            "value": "foo",
          },
        },
      ],
      "children": [],
      "component": false,
      "end": 21,
      "inline": false,
      "name": "div",
      "self_closing": false,
      "start": 0,
      "type": "Element",
    },
  ],
  "end": 21,
  "start": 0,
  "type": "Fragment",
}
`;

exports[`attribute > static unquoted 2`] = `
"let %%template0 = @html(\\"<div class=\\\\\\"foo\\\\\\"></div>\\");
let %fragment0 = @clone(%template0);
@append(%fragment0, $$root);
"
`;

exports[`component > v:self on a custom element 1`] = `
{
  "children": [
    {
      "attributes": [],
      "children": [
        {
          "attributes": [],
          "children": [
            {
              "decoded": "hello world!",
              "end": 27,
              "start": 15,
              "type": "Text",
              "value": "hello world!",
            },
          ],
          "component": true,
          "end": 36,
          "inline": true,
          "name": "v:self",
          "self_closing": false,
          "start": 7,
          "type": "Element",
        },
      ],
      "component": true,
      "end": 44,
      "inline": false,
      "name": "x-app",
      "self_closing": false,
      "start": 0,
      "type": "Element",
    },
  ],
  "end": 44,
  "start": 0,
  "type": "Fragment",
}
`;

exports[`component > v:self on a custom element 2`] = `
"let %%template0 = @html(\\"<x-app><div>hello world!</div></x-app>\\");
let %fragment0 = @clone(%template0);
let %child0 = new $$host.constructor();
let %marker0 = @traverse(%fragment0, [0, 0]);
@replace(%child0, %marker0, true);
@append(%fragment0, $$root);
"
`;

exports[`expression > component 1`] = `
{
  "children": [
    {
      "attributes": [],
      "children": [
        {
          "decoded": "hello ",
          "end": 12,
          "start": 6,
          "type": "Text",
          "value": "hello ",
        },
        {
          "end": 18,
          "expression": Node {
            "comments": [],
            "end": 17,
            "name": "name",
            "path": {
              "parent": undefined,
            },
            "start": 13,
            "trailingComments": [],
            "type": "Identifier",
          },
          "id": null,
          "start": 12,
          "type": "Expression",
        },
        {
          "decoded": "!",
          "end": 19,
          "start": 18,
          "type": "Text",
          "value": "!",
        },
      ],
      "component": true,
      "end": 26,
      "inline": true,
      "name": "Main",
      "self_closing": false,
      "start": 0,
      "type": "Element",
    },
  ],
  "end": 26,
  "start": 0,
  "type": "Fragment",
}
`;

exports[`expression > component 2`] = `
"let %%template0 = @html(\\"<div>hello <!>!</div>\\");
let %fragment0 = @clone(%template0);
let %marker0 = @traverse(%fragment0, [0, 1]);
@text(%marker0, () => name);
let %child0 = new Main();
let %marker1 = @traverse(%fragment0, [0]);
@replace(%child0, %marker1, true);
@append(%fragment0, $$root);
"
`;

exports[`expression > custom element 1`] = `
{
  "children": [
    {
      "attributes": [],
      "children": [
        {
          "decoded": "hello ",
          "end": 13,
          "start": 7,
          "type": "Text",
          "value": "hello ",
        },
        {
          "end": 19,
          "expression": Node {
            "comments": [],
            "end": 18,
            "name": "name",
            "path": {
              "parent": undefined,
            },
            "start": 14,
            "trailingComments": [],
            "type": "Identifier",
          },
          "id": null,
          "start": 13,
          "type": "Expression",
        },
        {
          "decoded": "!",
          "end": 20,
          "start": 19,
          "type": "Text",
          "value": "!",
        },
      ],
      "component": true,
      "end": 28,
      "inline": false,
      "name": "x-app",
      "self_closing": false,
      "start": 0,
      "type": "Element",
    },
  ],
  "end": 28,
  "start": 0,
  "type": "Fragment",
}
`;

exports[`expression > custom element 2`] = `
"let %%template0 = @html(\\"<x-app>hello <!>!</x-app>\\");
let %fragment0 = @clone(%template0);
let %marker0 = @traverse(%fragment0, [0, 1]);
@text(%marker0, () => name);
@append(%fragment0, $$root);
"
`;
